### ISSUE TITLE
Interpolation for the whole gird, or just active domain?

### ISSUE THREAD

**Issue by wkitlasten** [pypest/pyemu#277]
I am generating pilotpoint parameters for a fairly big model (400x500) with 15 layers. The active domain gradually reduces to about 500 cells by the last couple layers, so I end up with fewer pilot points in each layer (totally cool and expected). However, the "interp point loop" takes a while and appears to be interpolating for the entire grid (200,000 cells) for each layer. Is that expected? Would it be more efficient to interpolate only for the active domain... or just add unnecessary complexity?

My command:
```
for k in range(nlays):
    pf.add_parameters(filenames=k_file[k],par_type="pilotpoints",
                      use_pp_zones=True, zone_array=idom[k],
                      par_name_base=pargp, pargp=pargp,
                      upper_bound=10., lower_bound=0.1,
                      pp_space=4, geostruct=grid_gs)
```
Selected block of output:
```
2021-09-05 15:37:37.726603 finished: calculating factors for pargp=k_layer8_inst:1 took: 0:03:58.925442
2021-09-05 15:37:37.726603 starting: writing array-based template file 'c:\wairau\wairau_240_15_temp\k_layer8_inst1pp.dat.tpl'
2021-09-05 15:37:37.742236 saving zone array c:\wairau\wairau_240_15_temp\k_layer8_inst1pp.dat.zone for tpl file c:\wairau\wairau_240_15_temp\k_layer8_inst1pp.dat.tpl
2021-09-05 15:37:37.820810 finished: adding pilotpoints type multiplier style parameters for file(s) ['wairau_240_15.k_layer8.dat'] took: 0:04:01.335291
2021-09-05 15:37:37.852030 starting: adding pilotpoints type multiplier style parameters for file(s) ['wairau_240_15.k_layer9.dat']
2021-09-05 15:37:37.852030 starting: loading array c:\wairau\wairau_240_15_temp\wairau_240_15.k_layer9.dat
2021-09-05 15:37:38.005505 finished: loading array c:\wairau\wairau_240_15_temp\wairau_240_15.k_layer9.dat took: 0:00:00.153475
2021-09-05 15:37:38.005505 loaded array '..\zmodels\20210904_simulation\wairau_240_15_temp_pst_from\wairau_240_15.k_layer9.dat' of shape (415, 498)
2021-09-05 15:37:38.180174 starting: writing array-based template file 'c:\wairau\wairau_240_15_temp\k_layer9_inst1_pilotpoints.csv.tpl'
2021-09-05 15:37:38.180174 starting: setting up pilot point parameters
2021-09-05 15:37:38.180174 No spatial reference (containing cell spacing) passed.
2021-09-05 15:37:38.180174 OK - using spatial reference in parent object.
2021-09-05 15:37:39.654399 792 pilot point parameters created
2021-09-05 15:37:39.654399 pilot point 'pargp':k_layer9_inst:1
2021-09-05 15:37:39.654399 finished: setting up pilot point parameters took: 0:00:01.474225
2021-09-05 15:37:39.853038 starting: calculating factors for pargp=k_layer9_inst:1
2021-09-05 15:37:39.853038 saving krige variance file:c:\wairau\wairau_240_15_temp\k_layer9_inst1pp.var.dat
2021-09-05 15:37:39.853038 saving krige factors file:c:\wairau\wairau_240_15_temp\k_layer9_inst1pp.fac
starting interp point loop for 206670 points
starting 0
starting 1
starting 2
starting 3
starting 4
starting 5
starting 6
starting 7
starting 8
starting 9
took 240.6401 seconds
```

**Comment by jtwhite79** [comment]
We def only want to interpolate for the active cells.  Does the zone array have zeros for inactive cells?  Is the kriging solve time about the same for all layers (indicating we are solving for all cells rather than just active)?

**Comment by wkitlasten** [comment]
Yes. Yes? (+/-... for example: layer 1 = 235, layer 10 = 240, layer 15= 245)

For every layer it says: "starting interp point loop for 206670 points" 

layer 1 idomain:
![image](https://user-images.githubusercontent.com/14982084/132141529-2db2fd0b-a8ef-4a94-a926-e396e4c4beb9.png)

layer 10 idomain:
![image](https://user-images.githubusercontent.com/14982084/132141647-4dc470fa-c764-4acd-ac23-021b74848897.png)

layer 15 idomain:
![image](https://user-images.githubusercontent.com/14982084/132141762-90f3acd3-d8f9-4cd0-bbdb-dcf8862fe661.png)

**Comment by wkitlasten** [comment]
Some clues... around line 2295, 
if structured it calls ok_pp.calc_factors_grid with default num_threads=10
resulting in _calc_factors_mp() method, which doesn't handle nans in df.x and df.y

I tried changing num_threads=1 in ok_pp.calc_factors_grid call, but that is slow as!

I tried to dig into _calc_factor_mp() to deal with the nans, but I ain't that smart.

**Comment by briochh** [comment]
@jtwhite79, I think with the `_mp` version we are relying on the masking of "out of zone" xs and ys occurring already (e.g. in [`calc_factors_grid`](https://github.com/pypest/pyemu/blob/d69b8dc095c6606cb69e8ca0e57c94db4111c8a8/pyemu/utils/geostats.py#L987)). However, the shape of the df that gets set up in [L1370](https://github.com/pypest/pyemu/blob/d69b8dc095c6606cb69e8ca0e57c94db4111c8a8/pyemu/utils/geostats.py#L1370) (and the `_org` equivalent) are full length. So at least the reporting is all cells. Because the x and y are nan for those cells out of zones the factors will also be nan but the routine still need to cycle through them. With the _mp version it is hitting those manager list for each one which might be unnecessary overhead.

Can you think of any implications to dropping the Nans before the df construction? And truncating the `point_cov_data` in [L1375](https://github.com/pypest/pyemu/blob/d69b8dc095c6606cb69e8ca0e57c94db4111c8a8/pyemu/utils/geostats.py#L1375)? That way we are only looping in-zone cells?

**Comment by jtwhite79** [comment]
@briochh thanks checking this out.  The mp version shouldn't care if only only a subset of the grid points are being interpolated.  Can't see anything wrong with skipping...

**Comment by briochh** [comment]
@jtwhite79, nice one -- I am plumbing this in on a branch. We do need to keep hold of the full array for writing the var file, but this can be reconstructed after, I think. For @wkitlasten's test case (heaps of points) it reduces the the run time by... lots. Suggests that most of the `_mp` method slow down is in accessing those manager lists in the loop, not so much the actual calcs -- maybe it is in waiting for memory locks to open up.

**Comment by jtwhite79** [comment]
@briochh thats an interesting observation and honestly not that surprised.  maybe we need to tune the size of the pool...keen to see what kind of speed up we get from this!
